### PR TITLE
Implement syntactic merge checks for parameters.

### PR DESCRIPTION
**Pull request by jonmeow**
Note this isn't implementing checking through imports. The parse node there is harder to access through the context, so would require examining the entity in order to get the import declaration, to get at the ImportIR. We also don't have a parse tree attached in that case, and would need to add one to SemIR::File. But I believe we do want to add that, so it's explicitly a TODO.

Note GetTokenText re-lexes literal values, so there's a bit of potential overhead there. Not sure if we want a more efficient manner for comparing in cases like this.

### CODE REVIEW

**Comment by josh11b** [comment]
Please find another reviewer, as I'll be on vacation for the next two weeks.

File: toolchain/check/merge.cpp [hunk 1]
```diff
@@ -268,22 +268,99 @@ static auto CheckRedeclParams(Context& context, SemIRLoc new_decl_loc,
...
+// Returns true if the two nodes represent the same syntax.
+static auto IsNodeSyntaxEqual(Context& context, Parse::NodeId new_node_id,
+                              Parse::NodeId prev_node_id) -> bool {
+  if (context.parse_tree().node_kind(new_node_id) !=
+      context.parse_tree().node_kind(prev_node_id)) {
+    return false;
+  }
+
+  // TODO: Should there be a trivial way to check if we need to check spellings?
+  // Identifiers and literals need their text checked, but keywords and
```
Comment (zygoloid) [inline toolchain/check/merge.cpp]: (It'd be nice to just compare the `IdentifierId`s for identifiers where possible, though we'd need to check for `r#` somehow and still presumably need string comparisons for cross-file cases. But not in this PR!)

File: toolchain/check/merge.cpp [hunk 2]
```diff
@@ -268,22 +268,99 @@ static auto CheckRedeclParams(Context& context, SemIRLoc new_decl_loc,
...
+         "prev_first_param_node_id.is_valid";
+
+  auto new_range = context.parse_tree().postorder(new_first_param_node_id,
+                                                  new_last_param_node_id);
+  auto prev_range = context.parse_tree().postorder(prev_first_param_node_id,
+                                                   prev_last_param_node_id);
+
+  // zip is using the shortest range. If they differ in length, there should be
+  // some difference inside the range, or the semantics should have a mismatch
+  // (such as a different parameter count). As a consequence, we don't
```
Comment (zygoloid) [inline toolchain/check/merge.cpp]: I think we don't need to rely on a semantic mismatch here -- parse nodes are self-delimiting (we can reconstruct the tree structure from the sequence of nodes) so if the nodes match then they have the same tree structure, and because one of them is a complete tree, the other must be too.
```suggestion
  // some difference inside the range. As a consequence, we don't
```

**Review by zygoloid — APPROVED** [review]
Thanks!

Another edge case that would be useful to explicitly test (mostly because we've had discussions about using a different rule) is grouping parentheses differing between declarations.

File: toolchain/check/merge.cpp (hunk 2 above)
Comment (jonmeow) [reply toolchain/check/merge.cpp]: Adjusted, because I think it's important to note _why_ there should be a difference. Right now, that's actually because the `)` is part of the range.

File: toolchain/check/merge.cpp (hunk 1 above)
Comment (jonmeow) [reply toolchain/check/merge.cpp]: Adjusted

**Comment by jonmeow** [comment]
> Another edge case that would be useful to explicitly test (mostly because we've had discussions about using a different rule) is grouping parentheses differing between declarations.

Added test

Also, made me realize I wasn't testing deduced params explicitly, so added that too.

**Comment by jonmeow** [comment]
Per discussion, added a test for `r#` identifiers (because you'd mentioned them in a comment), and made them a TODO

**Review by zygoloid — APPROVED** [review]
